vortex-ipc/src/iterator.rs

```diff
@@ -0,0 +1,183 @@
+use vortex_buffer::Buffer;
+use vortex_dtype::DType;
+use vortex_error::{vortex_bail, vortex_err, VortexResult};
+
+use crate::messages::{DecoderMessage, EncoderMessage, MessageEncoder, SyncMessageReader};
+use crate::ALIGNMENT;
+
+/// An [`ArrayIterator`] for reading messages off an IPC stream.
+pub struct SyncIPCReader<R: Read> {
+    reader: SyncMessageReader<R>,
+    ctx: Arc<Context>,
+    dtype: DType,
+}
+
+impl<R: Read> SyncIPCReader<R> {
+    pub fn try_new(read: R, ctx: Arc<Context>) -> VortexResult<Self> {
+        let mut reader = SyncMessageReader::new(read);
+        match reader.read_message()? {
+            Some(msg) => match msg {
+                DecoderMessage::DType(dtype) => Ok(SyncIPCReader { reader, ctx, dtype }),
+                msg => {
+                    vortex_bail!("Expected DType message, got {:?}", msg);
+                }
+            },
+            None => vortex_bail!("Expected DType message, got EOF"),
+        }
+    }
+}
+
+impl<R: Read> ArrayIterator for SyncIPCReader<R> {
+    fn dtype(&self) -> &DType {
+        &self.dtype
+    }
+}
+
+impl<R: Read> Iterator for SyncIPCReader<R> {
+    type Item = VortexResult<ArrayData>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.reader.read_message().transpose()? {
+            Ok(msg) => match msg {
+                DecoderMessage::Array(array_parts) => Some(
+                    array_parts
+                        .into_array_data(self.ctx.clone(), self.dtype.clone())
+                        .and_then(|array| {
+                            if array.dtype() != self.dtype() {
+                                Err(vortex_err!(
+                                    "Array data type mismatch: expected {:?}, got {:?}",
+                                    self.dtype(),
+                                    array.dtype()
+                                ))
+                            } else {
+                                Ok(array)
+                            }
+                        }),
+                ),
+                msg => Some(Err(vortex_err!("Expected Array message, got {:?}", msg))),
+            },
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+/// A trait for converting an [`ArrayIterator`] into an IPC stream.
+pub trait ArrayIteratorIPC {
+    fn into_ipc(self) -> ArrayIteratorIPCBytes
+    where
+        Self: Sized;
+
+    fn write_ipc<W: Write>(self, write: W) -> VortexResult<W>
+    where
+        Self: Sized;
+}
+
+impl<I: ArrayIterator + 'static> ArrayIteratorIPC for I {
+    fn into_ipc(self) -> ArrayIteratorIPCBytes
+    where
+        Self: Sized,
+    {
+        let mut encoder = MessageEncoder::new(ALIGNMENT);
+        let buffers = encoder.encode(EncoderMessage::DType(self.dtype()));
+        ArrayIteratorIPCBytes {
+            inner: Box::new(self),
+            encoder,
+            buffers,
+        }
+    }
+
+    fn write_ipc<W: Write>(self, mut write: W) -> VortexResult<W>
```

Comment:
why is this here and not on `ArrayIteratorIPCBytes` directly?

Reply:
ArrayIteratorIPCBytes is an `Iterator<Item=VortexResult<Buffer>>`, I guess that could be written into a target? But typically iterator structs are single-purpose.